UI/window-basic-main.cpp

```diff
@@ -1648,6 +1666,364 @@ void OBSBasic::UpdateMultiviewProjectorMenu()
+		os_event_t     *complete_event; // event fired on done
+
+		int            stepIndex = 0; // step of the process
+
+		gs_stagesurf_t *stage_surf; // staging surface
+		bool           result = false; // function call result
+		std::string    outputFilePath; // output PNG file path
+
+		// reference to the callback function
+		void (*tick_callback_ref)(void *, float);
+	};
+
+	local_context callback_data;
+
+	// initialize completion event
+	os_event_init(&callback_data.complete_event, OS_EVENT_TYPE_AUTO);
+
+	// initialize callback function input
+	callback_data.outputFilePath = outputFilePath;
+	callback_data.width = video_info.base_width;
+	callback_data.height = video_info.base_height;
+
+	// Create stage surface
+	obs_enter_graphics();
+	callback_data.stage_surf = gs_stagesurface_create(
+			callback_data.width, callback_data.height,
+			GS_RGBA);
+	obs_leave_graphics();
+
+	// Set up our callback function body for
+	// obs_add_tick_callback()
+	callback_data.tick_callback_ref = [] (void *param,
+			float seconds)
+	{
+		local_context *args =
+			(local_context *)param;
+
+		// Worker function to be spawned after locking
+		// the graphics buffer.
+		auto save_worker_function = [args]() {
+			// Stage 3:
+			// Save acquired image data to PNG, release
+			// graphics buffer & staging surface, and
+			// signal the calling thread we are done.
+
+			args->result = SaveRGBABufferToPNGFile(
+					args->outputFilePath.c_str(),
+					args->width,
+					args->height,
+					args->image_data_ptr,
+					args->image_data_linesize);
+
+			// Release previously allocated gfx objects
+			obs_enter_graphics();
+			gs_stagesurface_unmap(args->stage_surf);
+			gs_stagesurface_destroy(args->stage_surf);
+			obs_leave_graphics();
+
+			// Signal the calling thread we're done
+			os_event_signal(args->complete_event);
+		};
+
+		switch (args->stepIndex) {
+		case 0:
+			// Stage 1:
+			// Copy the backbuffer to the stage surface
+			obs_enter_graphics();
+			gs_stage_texture(args->stage_surf,
+					obs_get_main_texture());
+			obs_leave_graphics();
+
+			// Increment our step number for the next
+			// call of the callback.
+			++args->stepIndex;
+			break;
+		case 1:
+			// Stage 2:
+			// Remove tick callback and map staging
+			// surface to a data pointer.
+			obs_remove_tick_callback(
+					args->tick_callback_ref, args);
+
+			// Lock the raw data to the image data buffer
+			obs_enter_graphics();
+			gs_stagesurface_map(args->stage_surf,
+					&args->image_data_ptr,
+					&args->image_data_linesize);
+			obs_leave_graphics();
+
+			// Spawn a thread to perform the save operation
+			std::thread save_thread = std::thread(
+					save_worker_function);
+			save_thread.detach();
+			break;
+		}
+	};
+
+	// Add tick callback
+	obs_add_tick_callback(callback_data.tick_callback_ref,
+			&callback_data);
+
+	// Wait for our callback to be fired and for video frame
+	// to become available
+	os_event_wait(callback_data.complete_event);
```

Comment:
You're blocking any callers of `SaveScreenshotToFile` for multiple frames, i.e. the UI thread will wait for multiple graphics thread frames; this really shouldn't happen imo, kind of the same problem as the initial design that blocked the graphics thread instead